Neo4j.Driver/Neo4j.Driver/Internal/Transaction.cs

```diff
@@ -90,27 +90,47 @@ protected override void Dispose(bool isDisposing)
                 {
                     if (_state == State.MarkedSuccess)
                     {
-                        _connection.Run(Commit, null, new BookmarkCollector(s => Bookmark = s));
-                        _connection.Sync();
-                        _state = State.Succeeded;
+                        try
+                        {
+                            _connection.Run(Commit, null, new BookmarkCollector(s => Bookmark = s));
+                            _connection.Sync();
+                            _state = State.Succeeded;
+                        }
+                        catch(Exception)
+                        {
+                            // if we ever failed to commit, then we rollback the tx
+                            try
+                            {
+                                RollBackTx();
+                            }
+                            catch (Exception)
+                            {
+                                // ignored
+                            }
+
+                            throw;
+                        }
                     }
                     else if (_state == State.MarkedFailed || _state == State.Active)
                     {
-                        // If alwaysValid of the things we've put in the queue have been sent off, there is no need to
-                        // do this, we could just clear the queue. Future optimization.
-                        _connection.Run(Rollback, null, new BookmarkCollector(s => Bookmark = s));
-                        _connection.Sync();
-                        _state = State.RolledBack;
+                        RollBackTx();
                     }
                 }
             }
             finally
             {
-                _cleanupAction.Invoke();;
+                _cleanupAction.Invoke();
                 base.Dispose(true);
             }
         }
 
+        private void RollBackTx()
+        {
+            _connection.Run(Rollback, null, new BookmarkCollector(s => Bookmark = s));
```

Comment:
Oh, that makes sense. Similarly here, we should do `Collector.NO_OP`
 https://github.com/neo4j/neo4j-java-driver/blob/1.1/driver/src/main/java/org/neo4j/driver/internal/ExplicitTransaction.java#L161